include/deal.II/grid/tria_accessor.templates.h

```diff
@@ -2040,6 +2040,163 @@ TriaAccessor<structdim, dim, spacedim>::diameter () const
+      const Point<spacedim> p34( this->vertex(3)-this->vertex(4));
+      const std::vector<double> diagonals= { p70.norm(),
+                                             p61.norm(),
+                                             p25.norm(),
+                                             p34.norm()
+                                           };
+      const std::vector<double>::const_iterator
+      it = std::max_element( diagonals.begin(), diagonals.end());
+      if (it == diagonals.begin())
+        {
+          center = this->vertex(0) + 0.5* p70;
+          vertices_within_ball[7] = true;
+          vertices_within_ball[0] = true;
+        }
+      else if (it == diagonals.begin()+1)
+        {
+          center = this->vertex(1) + 0.5* p61;
+          vertices_within_ball[6] = true;
+          vertices_within_ball[1] = true;
+        }
+      else if (it == diagonals.begin()+2)
+        {
+          center = this->vertex(5) + 0.5* p25;
+          vertices_within_ball[2] = true;
+          vertices_within_ball[5] = true;
+        }
+      else
+        {
+          center = this->vertex(4) + 0.5* p34;
+          vertices_within_ball[3] = true;
+          vertices_within_ball[4] = true;
+        }
+      radius = *it * 0.5;
+      break;
+    }
+    default:
+      Assert (false, ExcNotImplemented());
+      return std::pair<Point<spacedim>,double>();
+    }
+
+  // For each vertex that is found to be geometrically outside the ball
+  // enlarge the ball  so that the new ball contains both the previous ball
+  // and the given vertex.
+  for (unsigned int v = 0; v < GeometryInfo<structdim>::vertices_per_cell; ++v)
+    if (!vertices_within_ball[v])
+      {
+        const double distance = center.distance(this->vertex(v));
+        if (distance > radius)
+          {
+            // we found a vertex which is outside of the ball
+            // extend it (move center and change radius)
+            const Point<spacedim> pCV (center - this->vertex(v));
+            radius = (distance + radius) * 0.5;
+            center = this->vertex(v) + pCV * (radius / distance);
+
+            // Now the new ball constructed in this block
+            // encloses the vertex (v) that was found to be geometrically
+            // outside the old ball. Since the ball is enlarged to enclose
+            // the vertex (v) we could update it's vertex mask to true.
+            // But we don't need to.
+            // vertices_within_ball[v] = true;
+          }
+        else
+          {
+            // Vertex is found to be geometrically inside the ball
+            // but it's corresponding entry in the vertex mask isn't updated.
+            // We could update the vertex mask but we don't need to.
+            // vertices_within_ball[v] = true;
+          }
+      }
+
+  // Ensure that all the vertices are within the ball if run in debug mode
+  Assert (check_enclosing_ball_internal(*this, center, radius),
+          ExcInternalError());
```

Comment:
@masterleinad :
i asked @vishalkenchan to do it this way as I though that's how we normally do it for complicated checks. 

Do you suggest to write this as
```
#ifdef DEBUG
    bool res = true;
    // for()

    Assert(res, ....);
#endif
```